### PR TITLE
[dbnode] Entry reconciliation for batched writes

### PR DESCRIPTION
**Pull request by arnikola**
Batched writes are vulnerable to issues from stale index state, where an
update to mark an indexed entry as successful may be written to a stale
index entry. This PR adds reconciliation for the entry prior to marking
it as done, ensuring that the live entry is marked as done, as well as
 metrics tracking when reconciliation is required.

Related PRs:
https://github.com/m3db/m3/pull/3905/files fixes reads with stale state

**What this PR does / why we need it**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note

```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note

```

### CODE REVIEW

**Review by ryanhall07 — APPROVED** [review]
Changes look fine. I'm worried about this being the third time we've had to fix a stale index entry. Have we at least done a code search for all potential affected write paths?

File: src/dbnode/storage/index/types.go
Comment (rallen090) [inline src/dbnode/storage/index/types.go:844]: The only worry here is the added shard map lock contention that this reconciliation costs. Before we land this we can test in some env at scale and get a sense of any perf impact.

**Review by rallen090 — APPROVED** [review]
LGTM pending perf testing

**Review by nbroyles — APPROVED** [review]

File: src/dbnode/storage/index/types.go
Comment (robskillington) [inline src/dbnode/storage/index/types.go:181]:
```suggestion
	// mutable and the most recent need to shadow older entries.
```

File: src/m3ninx/doc/types.go
Comment (robskillington) [inline src/m3ninx/doc/types.go:149]: nit: Getter names don't need "Get" prefix as per effective Go?

File: src/m3ninx/doc/types.go
Comment (robskillington) [inline src/m3ninx/doc/types.go:87]: nit: Getter names don't need "Get" prefix as per effective Go, maybe just "IDString"?

File: src/dbnode/storage/index/types.go
Comment (robskillington) [reply src/dbnode/storage/index/types.go:849]: There's also the possibility that this is the entry that will be orphaned and that there is nothing in the shard map at all right now (in fact this is more likely since orphans are typically a side effect of a first insertion which doesn't win the race to be inserted into the shard map).

Instead of adding potential performance overhead to the write path, couldn't we do this same merge of states logic when we consider it for garbage collection and/or read reconciliation time?
https://github.com/m3db/m3/blob/82aa22e67816fe2dcf08c50c6380bfac8ce9a450/src/dbnode/storage/index/mutable_segments.go#L233-L247

I assume that's why we're doing this at all for higher accuracy in both read reconcilation and garbage collection, so wouldn't it be better to pay the cost in those parts of the code base?

Most things are architected so that write path stays lean which causes complexity downstream, but this seems to go against that pattern somewhat so I just want to understand if we can achieve the same thing but in code paths where we're willing to pay more cycles/latency (which for the background garbage collection path of the index we have plenty more cycles/latency to spare).

File: src/dbnode/storage/entry.go
Comment (rallen090) [inline src/dbnode/storage/entry.go:183]: Is there a reason for this `Get`? IMO we can just have a single merge func that takes in the other entry.

Comment (rallen090) [reply src/dbnode/storage/entry.go:183]: Also I think we can avoid this map copy alloc and just mutate the entry's map directly that we call merge on

File: src/m3ninx/index/segment/builder/multi_segments_builder.go
Comment (robskillington) [inline src/m3ninx/index/segment/builder/multi_segments_builder.go:156]: I see yeah we kind of want to do patch up state if we filter out a duplicate.

I would probably update the DocumentsFilter interface to be three methods:
```go
type DocumentsFilter interface {
  ContainsDoc(d doc.Metadata) bool
  OnNotContainsDoc(d doc.Metadata)
  OnDuplicateDoc(d doc.Metadata)
}
```

That way we don't have to pass three different distinct things to the multi segments builder (the doc filter, the filter count and the reconcile duplicate function).

Then this becomes:
```go
		for iter.Next() {
			d := iter.Current()
			negativeOffsets = append(negativeOffsets, currOffset)
			if b.idSet.Contains(d.ID) {
				// Skip duplicates.
				negativeOffsets[len(negativeOffsets)-1] = -1
				currOffset++
				if b.filter != nil {
					b.filter.OnDuplicateDoc(d) // Callback here to play TryReconcileDuplicates
				}
				continue
			}
			if b.filter != nil && !b.filter.ContainsDoc(d) { // Existing call here to determine if should GC or not
				// Actively filtering and ID is not contained.
				negativeOffsets[len(negativeOffsets)-1] = -1
				currOffset++
				b.filter.OnNotContainsDoc(d) // Callback here to increment filter count
				continue
			}
			b.idSet.SetUnsafe(d.ID, struct{}{}, IDsMapSetUnsafeOptions{
				NoCopyKey:     true,
				NoFinalizeKey: true,
			})
			b.docs = append(b.docs, d)
			added++
		}
```

File: src/dbnode/storage/entry.go
Comment (rallen090) [inline src/dbnode/storage/entry.go:319]: Should we put the read locking within the Merge func?

File: src/m3ninx/index/segment/builder/multi_segments_builder.go
Comment (rallen090) [reply src/m3ninx/index/segment/builder/multi_segments_builder.go:156]: What's the reason for passing in these functions to mutableSegments at all? IMO it might just be more clear for mutableSegments to call `d.Metadata.OnIndexSeries.Merge(...)` directly?

File: src/dbnode/storage/entry.go
Comment (arnikola) [reply src/dbnode/storage/entry.go:319]: So the merge actually runs on the reconciled entry; we need to take a read lock here though since we're touching `entry.reverseIndex.states` on the entry itself to feed into the reconciled one

File: src/m3ninx/index/segment/builder/multi_segments_builder.go
Comment (arnikola) [reply src/m3ninx/index/segment/builder/multi_segments_builder.go:156]: Nice! I would be leaning towards Ryan's approach here; think the main upside of this `DocumentsFilter` approach would be getting metrics plumbed through, but we already have some through the `Entry` plumbing that's part of this PR; happy to refactor it to lean more heavily on the `DocumentsFilter` approach though if that can give us some benefit. Also needs a little annoying refactoring here: https://github.com/m3db/m3/blob/master/src/dbnode/storage/index/mutable_segments.go#L771 since we'd want to set a "partial" `DocumentsFilter` (not a huge change but still a bit annoying)

File: src/dbnode/storage/entry.go
Comment (arnikola) [reply src/dbnode/storage/entry.go:183]: Yeah removed this after the refactor to move to the segments builder:+1:

File: src/m3ninx/doc/types.go
Comment (arnikola) [reply src/m3ninx/doc/types.go:149]: Managed to refactor this out 👍

File: src/dbnode/storage/index/types.go
Comment (arnikola) [reply src/dbnode/storage/index/types.go:181]: Oops heh

File: src/dbnode/storage/entry.go
Comment (rallen090) [reply src/dbnode/storage/entry.go:360]: Ah I see - good call

File: src/dbnode/storage/entry.go
Comment (rallen090) [inline src/dbnode/storage/entry.go:396]: nit: can probably share this logic w/ TryMarkGC w/ a private func - just because this pattern of retrieving series, dec, and merge, are all in both places.

File: src/m3ninx/index/segment/builder/multi_segments_builder.go
Comment (rallen090) [inline src/m3ninx/index/segment/builder/multi_segments_builder.go:154]: nit: maybe more detailed comment here on why it is important to reconcile these (i.e. primarily merge the time ranges)

**Review by rallen090 — APPROVED** [review]

File: src/dbnode/storage/entry.go
Comment (arnikola) [reply src/dbnode/storage/entry.go:396]: It's a bit annoying since the dec is a `defer` so we can't run that from a private function, and because of that it functionally ends up having a fairly similar amount of boilerplate either way; may leave it as is

**Pull request merged**
